lib/storage/metricsmetadata/store.go

```diff
@@ -0,0 +1,382 @@
+package metricsmetadata
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/bytesutil"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fasttime"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/fs"
+	"github.com/VictoriaMetrics/VictoriaMetrics/lib/logger"
+)
+
+type metadataKey struct {
+	accountID        uint32
+	projectID        uint32
+	metricFamilyName string
+}
+
+type metricTimingInfo struct {
+	LastIngestionTime    uint64 `json:"lastIngestionTime"`
+	AvgIngestionInterval uint64 `json:"avgIngestionInterval"`
+	IngestionCount       int64  `json:"ingestionCount"`
+}
+
+type recordForStore struct {
+	AccountID        uint32
+	ProjectID        uint32
+	MetricFamilyName string
+	Rows             []Row
+	TimingInfo       metricTimingInfo
+}
+
+type MetadataStoreMetrics struct {
+	ItemsTotal         uint64
+	ItemsIngestedTotal uint64
+	ItemsDeduplicated  uint64
+	ItemsDeleted       uint64
+	ItemsSizeBytes     uint64
+}
+
+type Store struct {
+	MetricsMetadataStorage    map[metadataKey][]Row
+	MetricTimingInfo          map[metadataKey]metricTimingInfo
+	metricMetadataStorageLock sync.RWMutex
+
+	storagePath string
+
+	itemsIngestedTotal uint64
+	itemsDeduplicated  uint64
+	itemsDeleted       uint64
+	itemsCurrentTotal  uint64
+
+	cleanupInterval time.Duration
+	cleanupStopCh   chan struct{}
+	cleanupWG       sync.WaitGroup
+}
+
+func NewStore() *Store {
+	s := &Store{
+		MetricsMetadataStorage: make(map[metadataKey][]Row),
+		MetricTimingInfo:       make(map[metadataKey]metricTimingInfo),
+		cleanupInterval:        5 * time.Minute,
+		cleanupStopCh:          make(chan struct{}),
+	}
+
+	s.cleanupWG.Add(1)
+	go s.runCleanupScheduler()
+
+	return s
+}
+
+func MustLoadFrom(path string) *Store {
+	s := NewStore()
+	err := s.LoadFrom(path)
+	if err != nil {
+		logger.Panicf("cannot load metrics metadata from %q: %s", path, err)
+	}
+	return s
+}
+
+func (s *Store) MustClose() {
+	close(s.cleanupStopCh)
+	s.cleanupWG.Wait()
+	s.metricMetadataStorageLock.Lock()
+	s.mustSaveLocked()
+	s.metricMetadataStorageLock.Unlock()
+}
+
+func (s *Store) Add(rows []Row) error {
+	if len(rows) == 0 {
+		return nil
+	}
+
+	s.metricMetadataStorageLock.Lock()
+	defer s.metricMetadataStorageLock.Unlock()
+
+	// Update timing for all metrics that were touched
+	now := fasttime.UnixTimestamp()
+
+	for _, mr := range rows {
+		key := metadataKey{
+			accountID:        mr.AccountID,
+			projectID:        mr.ProjectID,
+			metricFamilyName: bytesutil.ToUnsafeString(mr.MetricFamilyName),
+		}
+		s.updateMetricTimingLocked(key, now)
+
+		metadataRows, ok := s.MetricsMetadataStorage[key]
+		if !ok {
+			s.MetricsMetadataStorage[key] = make([]Row, 0, 1)
```

Comment:
shouldn't we copy key in this case? seems like `rows` can be reused in mrUnmarshalWork